Integrates H2O Driverless AI MOJO Scoring Pipeline into Apache NiFi

james94 opened the pull request:
This pull request resolves GitHub filed issue: Create [H2O Driverless AI MOJO Deployment to Apache NiFi](https://github.com/h2oai/dai-deployment-examples/issues/17) #17

**NiFi and H2O Driverless AI Integration** via Custom NiFi Processor:

Integrates NiFi with H2O Driverless AI by using Driverless AI's MOJO Scoring Pipeline (in Java Runtime) and NiFi's Custom Processor. This processor executes the MOJO Scoring Pipeline to do batch scoring or real-time scoring for one or more predicted labels on tabular data in the incoming flow file content. If the tabular data is one row, then the MOJO does real-time scoring. If the tabular data is multiple rows, then the MOJO does batch scoring. I would like to contribute the **mojo-nifi** as a new deployment example, which could be transitioned to a production ready deployment template.

**1 Custom Processor** created for NiFi:

**ExecuteDaiMojoScoringPipeline** - Executes H2O Driverless AI's MOJO Scoring Pipeline in Java Runtime to do batch scoring or real-time scoring on a frame of data within each incoming flow file. It requires the user to add **mojo2-runtime.jar** filepath into **MOJO2 Runtime JAR Directory** property to dynamically modify the classpath. It also requires the user to add the **pipeline.mojo** filepath into the **Pipeline MOJO Filepath** property. This property is used in the onTrigger() method to get the pipeline.mojo filepath, so we can pass it into the
MojoPipeline.loadFrom(pipelineMojoPath) to instantiate our MojoPipeline model. Then the record read in with Record Reader and the model are passed into a predict() method to make predictions on the test data within the record. Inside the predict() method, I use MojoFrameBuilder and MojoRowBuilder with the recordMap to build an input MojoFrame. Then I use the model's transform(input MojoFrame) method to make the predictions on the input and store them into an output MojoFrame. I iterate through the MojoFrame by row and column to store each key value pair prediction into the predictedRecordMap. I then convert the predictedRecordMap to predictedRecord and return the record back to onTrigger to write the record to the flow file content using RecordSetWriter. We keep writing predicted Records to the flow file content until there are no more records to write. Then we reach near end of onTrigger() and the flow file is either transferred on relationship failure, success or original to the next connection.

**Hydraulic System Condition Monitoring** Data used in NiFi Flow:
 
The sensor test data I used in this integration comes from UCI ML Repo: Condition Monitoring for Hydraulic Systems. I was able to predict the hydraulic cooling condition through NiFi and H2O Integration described above. This use case is hydraulic system predictive maintenance.

james94 committed:
Added link to NiFi MOJO Deployment Example

james94 committed:
Added .gitignore for nifi dai mojo processor

james94 committed:
Added how to use the NiFi processor to execute MOJO

I include steps on how to use the NiFi custom processor to
execute the MOJO scoring pipelien within a NiFi Data Flow
to do batch scoring or real-time scoring.

james94 committed:
Updated the description for NiFi deployment example

james94 committed:
Added images for how to use NiFi MOJO processor

james94 committed:
Added NiFi custom processor project

This project includes the NiFi processor Java code for
executing the Driverless AI MOJO Scoring Pipeline using
the NiFi processor. It also includes a Junit test program
for testing different properties and test cases of the
processor. I also included two NiFi flow templates on how
I used the NiFi ExecuteDaiMojoScoringPipeline processor
on Hydraulic System sensor data to predict Hydraulic
cooling condition. One NiFi templates pulls in real-time
data, so we do real-time scoring. The other NiFi template
pulls in batch data, so we do batch scoring

james94 committed:
Added nifi model deployment common folder

The NiFi MOJO model deployment common folder makes it really easy to use
the NiFi custom processor ExecuteDaiMojoScoringPipeline in a NiFi flow
because the folder/ comes with the mojo-pipeline folder, testData folder
and the mojo2-runtime.jar

james94 committed:
Tell git to ignore predictions in predData

james94 committed:
Added onScheduled() method, so MojoPipeline loaded once

Added NiFi onSchedule(), so MojoPipeline only gets loaded once
for each time the processor is scheduled to run. Earlier when
the MojoPipeline was instantiated in onTrigger() method, the
MojoPipeline would get loaded every time data was scored. Now
the MojoPipeline only gets instantiated once, then the MojoPipeline
gets executed every time data is scored.

james94 committed:
Delete predData

nkpng2k commented on mojo-nifi/README.md:
```diff
@@ -0,0 +1,160 @@
+
+## Prerequisites
+
+- Java >= 7 runtime (JDK >= 1.7)
```
nit: I'd be inclined to just say java 8 since thats more or less what we support.

james94 replied:
@nkpng2k The Driverless AI MOJO Scoring Pipeline documentation suggests Java >= 7 runtime (JDK >= 1.7)

http://docs.h2o.ai/driverless-ai/1-8-lts/docs/userguide/scoring-mojo-scoring-pipeline.html

nkpng2k commented on mojo-nifi/README.md:
```diff
@@ -0,0 +1,160 @@
+    - comes with this repo under `model-deployment/common/hydraulic/testData/`
+
+- MOJO2 Runtime JAR
+    - comes with this repo under `model-deployment/common/mojo2-runtime/java-runtime/java/mojo2-runtime.jar`
```
this should not be uploaded. too hard to maintain. point properly to mojo runtime in maven/gradle deps.

james94 replied:
@nkpng2k I will remove the uploaded MOJO2 Runtime JAR. I will just mention that the user should download the MOJO Scoring Pipeline. I also got advice from an SE customer perspective not to point to mojo runtime in maven/gradle deps and instead have the user provide the MOJO2 Runtime JAR that they get when they download the MOJO Scoring Pipeline. So, in maven I mention that the user will provide the MOJO2 Runtime JAR instead of having maven auto download it from the internet. I heard from an SE that some of our customers in the hospital setting wouldn't risk the security vulnerability.

nkpng2k replied:
you can always override compiled jar with `-cp` to include mojo runtime provided by dai. For example: `java -cp mojo2-runtime.jar:apachi-nifi-integration.jar` then whatever is compiled in to the `apache-nifi-integration.jar` is overlooked. Behavior of java classloader. 

proper build process is a good practice to adhere too.

james94 replied:
@nkpng2k I can update the place in the maven xml code in which I let maven know the user will supply the path to the MOJO2 Runtime JAR and instead tell maven to download the latest MOJO2 Runtime JAR for dai. Currently, I have the user pass in the filepath of their MOJO2 Runtime JAR to the NiFi processor property on the NiFi UI. I can edit the Java code, so that it is optional for the user to pass their MOJO2 Runtime JAR filepath to the NiFi processor on the NiFi UI.

james94 replied:
@nkpng2k I updated the maven pom.xml, so maven downloads the mojo2-runtime api. Now the user will not need to provide their mojo2-runtime.jar file to the processor. Since I built the experiment and downloaded the MOJO SP from Driverless AI 1.8.7.1, I tell maven to download mojo2-runtime API version 2.4.2.

nkpng2k commented on mojo-nifi/README.md:
```diff
@@ -0,0 +1,160 @@
+
+~~~bash
+# Linux user
+echo "export DRIVERLESS_AI_LICENSE_KEY={license_key}" >> ~/.profile
```
this could be dangerous... but maybe just me. I'd advise just setting this one time `export ....` not save to actual profile.

james94 replied:
@nkpng2k I will change this line to set the license key as a temporary environment variable.

nkpng2k commented on mojo-nifi/README.md:
```diff
@@ -0,0 +1,160 @@
+
+### Process For Developing the Processor
+
+For developing the NiFi custom processor, I used Maven and Eclipse IDE. When using NiFi with Maven, there is the option to build a starter template for the custom processor you are going to build. To generate the starter template, I used the command: `mvn archetype:generate`. Next my plan was to work on the processor's code in Eclipse. I ran the command `mvn install` to compile the project. Then  ...
```
nit: good to have but its a readme. this may not be necessary. I'd avoid `I` statements personally. Only add things that will make things easier for the devs/users in the future. IMHO

james94 replied:
@nkpng2k I removed this section of the README and updated the rest of it to focus on just going through the steps to deploy the dai MOJO to NiFi.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/resources/META-INF/services/org.apache.nifi.processor.Processor:
```diff
@@ -0,0 +1,15 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+org.apache.nifi.processors.h2o.record.ExecuteDaiMojoScoringPipeline
```
nit: new lines at end of file please.

james94 replied:
@nkpng2k All the files have a new line at the end.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/test/java/org/apache/nifi/processors/h2o/record/TestExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,303 @@
+	
+	// The pretty printed json comparisons don't work on windows
+	@BeforeClass
+	public static void setUpSuit() {
```
typo: `setUpSuite`

james94 replied:
@nkpng2k this typo has been addressed.

nkpng2k commented on mojo-nifi/model-deployment/common/hydraulic/mojo-pipeline/0_LightGBMModel_0.dot:
```diff
@@ -0,0 +1,48 @@
+digraph {
```
we probably don't need all these extra files? I think these are all just for visualizations.

james94 replied:
@nkpng2k I removed the mojo-pipeline/ folder from the GitHub repo, so now this file and all others that came with the folder are gone.

nkpng2k commented on mojo-nifi/model-deployment/common/hydraulic/testData/test-real-time-data/test_1.csv:
```diff
@@ -0,0 +1 @@
+155.64058,104.911064,0.8626985,0.000211,8.370247,8.327606,2.0297766,8.869429,35.326817,40.8748,38.30345,30.47345,2367.348,0.5243667,1.3104666,29.127466
```
nit: I'd actually recommend against having all this data stored in github. The alternative would be maybe a small python/bash script to obtain the data from say s3 and then break it into the smaller bits.

james94 replied:
@nkpng2k I removed the testData from the GitHub repo.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/templates/predBatchesHydCoolCond.xml:
```diff
@@ -0,0 +1,1129 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
1.2k lines diff? what are these please?

james94 replied:
@nkpng2k these files are NiFi Data Flow templates, which are prebuilt NiFi Data Flows that have the custom MOJO NiFi processor.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,377 @@
+			rowBuilder.setValue(recordEntry.getKey(), String.valueOf(recordEntry.getValue()));
+		}
+		// Append a row from the current state of the rowBuilder
+		frameBuilder.addRow(rowBuilder);
```
:thinking: this seems to only ever score a single line at a time. What would happen if there is more than one record in the score request? is that even possible?

james94 replied:
@nkpng2k earlier in the `onTrigger()` method, let's say batch data comes into the NiFi processor. I use a NiFi FlowFile object called `original`. Then a NiFi RecordReader object named `reader` is created to read the FlowFile named `original`. With `reader` I am able to read the FlowFile `original` one row at a time using `nextRecord()` method and that record gets stored into `firstRecord`, then `predict()` is performed. If there is no exception thrown, then the next record is stored into `record` and `predict()` is performed. This process repeats until all records are scored.

There will only be 1 line (or 1 row) per NiFi record. The `predict()` method only executes on one record at a time. There is only ever one record at a time per score request.

james94 replied:
@nkpng2k I can think of a case where there may be more than one record in the score request that comes into this custom NiFi processor. Imagine you have two or more NiFi processors pulling in data (let's say Hydraulic System Data) and these NiFi processors connect to the custom processor, then the custom processor will perform predictions on that data. Which connection transmitting the data does the custom processor perform predictions on first? That depends on which data arrives to the processor first and how the custom processor's concurrent tasks setting is configured. The  processor's concurrent task controls how many threads the Processor will use, which controls how many Flow Files (the data contents and meta data attributes) should be processed by this custom processor at the same time.

Reference based on NiFi Developer's Answer: https://stackoverflow.com/questions/46553181/nifi-parallel-and-concurrent-execution-with-executestreamcommand

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,377 @@
+	}
+	
+	@OnScheduled
+	public void onScheduled(final ProcessContext context) {
```
qq - does this load the mojo every time a new record is scheduled to be scored? or is this loading only one time. I would recommend better to load the mojo only once and retain it in memory as class attribute.

james94 replied:
@nkpng2k this onScheduled() method loads the MOJO only once and retains it in memory as a class attribute. Then the NiFi processor is able to make predictions on the data as it comes into the processor.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,377 @@
+		try {
+			model = MojoPipeline.loadFrom(pipelineMojoPath);
+		} catch (Exception e) {
+			e.printStackTrace();
```
nit: maybe `log4j` for proper logging? I don't understand the behavior of nifi, but generally I'd advise against just dumping stacks to stderr/stdout

james94 replied:
@nkpng2k I replaced the e.printStackTrace() with Java getLogger("...") for error logging.

james94 replied:
@nkpng2k I removed try catch block in the onScheduled method and used throws to indicate that an IOException or LicenseException can be thrown from this method stopping the program.

james94 replied:
Originally I was looking at ending the program if the `pipeline.mojo` file was not found, but instead I found NiFi comes with standard validators for properties. Since this property `PIPELINE_MOJO_FILEPATH` is a path to a file, I added a validator to verify that the file exists using `.addValidator(StandardValidators.FILE_EXISTS_VALIDATOR)`. So, if a user enters the wrong path to the `pipeline.mojo`, NiFi UI will show an alert on the top left side of the processor indicating that the file does not exist.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,377 @@
+		final List<ValidationResult> results = new ArrayList<>(super.customValidate(validationContext));
+		final String pipelineMojoPath = validationContext.getProperty(PIPELINE_MOJO_FILEPATH).isSet() ? validationContext.getProperty(PIPELINE_MOJO_FILEPATH).getValue() : null;
+		
+		if(pipelineMojoPath == null) {
```
qq - I know I sort of saw this in the demo, but how does the mojo actually get put to the runtime environment? `scp`? What happens if the user does not have `ssh` access to the machine they are running this on?

james94 replied:
@nkpng2k for the demo I did, the **pipeline.mojo** gets put to this runtime environment via `scp`. If the user doesn't have `ssh` access to the machine, they would not be able to get the MOJO. However, if we go beyond the demo I built, we could build a separate NiFi Data Flow that every so often will download the latest MOJO Scoring Pipeline from Driverless AI to the machine NiFi is running on using the Driverless AI Python Client.

nkpng2k replied:
ok, I'd not be surprised if the people trying to run this do not have access to the box that they are trying to run this on, but probably is a task for another date.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,377 @@
+					attributes.putAll(writeResult.getAttributes());
+				}
+				
+				final String mojoPipelineUUID = "pipeline.mojo uuid " + model.getUuid();
```
hmm this is a weird place to put this... after the scoring attempt has been made?

james94 replied:
@nkpng2k I moved the mojoPipelineUUID before the scoring takes place. Now it is on line 236.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,377 @@
+				    ) {
+					writer.beginRecordSet();
+					
+					writer.write(scoredFirstRecord);
```
why is there 2 different calls of `predict` one for the first record and then everything else?

james94 replied:
@nkpng2k the first call to `predict` is performed, then a check is done to make sure there wasn't an error scoring the first record before proceeding onward with making all the other predictions. 

- If an error did happen when scoring the first record, a ProcessException is thrown to stop all execution after the throw statement. So, the predict method is not executed on all remaining records. The program resumes execution when the exception is caught by the catch block later in the program.
- Else if no error occurred on scoring the first record, the first scored record is used to write the schema for the output record prediction, then the predict method is executed on all remaining records and that schema will be applied to all remaining scored records.

nkpng2k replied:
hmmm maybe I do not understand enough the nifi logic, but this seems overly complicated. Why not a single try/catch clause. Just because the first record scored successfully does not mean that there will not be a bad row elsewhere in the data. what happens then? Also, we should already know the schema ahead of time, we should not need to score a row in order to determine schema of input and output. seems like chicken before the egg.

james94 replied:
@nkpng2k You have a great point. I removed the first call to `predict()`. In the `onScheduled()`, after the MojoPipeline model is loaded, I use the model's `getOutputMeta()` to get the schema for predictions, then I convert the MojoFrame Meta Data to a NiFi Record Schema since NiFi doesn't understand MojoFrame Meta Data. This process is only done 1 time after the user turns on this NiFi processor. Then later in the `onTrigger()`, I use the scoredMojoRecordSchema that we got from the model to create the scored writeSchema. Then while there are still records to perform predictions on, the `predict()` is executed. 

Thank you for sharing your suggestion, the previous way where I was using 2 different calls of `predict()` was doing extra work and this updated approach saves us some execution time and is less complicated.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,377 @@
+					
+					writeResult = writer.finishRecordSet();
+
+					try {
```
`finally`

james94 replied:
@nkpng2k I removed the try catch statements from line 263 to 267 because RecordSetWriter has multilevel inheritance to AutoCloseable, so since RecordSetWriter is instantiated as a try resource, once execution leaves the try resource block, the RecordSetWriter is automatically closed.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,377 @@
+		Map<String, Object> predictedRecordMap = new HashMap<>();
+		
+		// loop on the number of rows of each column in the MojoFrame
+		for(int row_i = 0; row_i < oframe.getNrows(); row_i++)
```
if there is only ever one row in the record then this seems unnecessary?

james94 replied:
@nkpng2k there is only one row in each record, so I can remove the first loop. I will add this update to the code.

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,377 @@
+	}
+	
+	@SuppressWarnings("unchecked")
+	private Record predict(final Record record, MojoPipeline model, final ComponentLog logger) {
```
this method does not seem to be doing just prediction. It seems to be translating records to mojo frames among other things. in fact I would recommend a separated class for this. like `recordHandler` which does the translations between records and mojos. or like a `MojoTransform` class that has multiple methods each doing its own special task.

james94 replied:
@nkpng2k An alternative to a separate class could be to break this method into smaller methods within the already existing class that perform certain tasks, such as translating the records to mojo frames and other tasks. But I do think a separate class would be cleaner. I will try breaking this method into separate methods and see how that looks. If it is too much for the current class, then I will create a separate class for `MojoTransform`. I may call the separate class a different name.

nkpng2k replied:
this is fine, I just strongly advocate for properly named methods and classes. It helps for development later.

james94 replied:
@nkpng2k I broke apart the predict() method, created a convertRecordToMojoFrame(...) method and a convertMojoFrameToRecord(...) method. 

If we look at the flow when the NiFi processor is going to execute the MOJO to perform scoring we see the following happen:
- the NiFi record gets converted to a MojoFrame via convertRecordToMojoFrame(...)
- the MOJO SP model performs scoring on the MojoFrame via predict()
- the scored MojoFrame gets converted to a NiFi scored Record via convertMojoFrameToRecord(...)

nkpng2k reviewed (dismissed):
1 global thing: there are 2 mojo runtime jars in the pr, I would advise not having them here at all. Use maven to manage the dependencies. 

Couple different comments. Some may be irrelevant since I don't understand NiFi behavior.

james94 committed:
Updates added based on nkpng2k's Feedback

In prerequisites section of mojo-nifi/ README, removed Java runtime since in the walkthrough
I go through the instruction to install it.

In prerequisites section narrowed it down to suggest user has a Driverless AI environment,
an Ubuntu 18.04 EC2 instance and cloned the dai-deployment-examples repo on their local
machine.

Instead of having the MOJO2 Runtime JAR uploaded to GitHub, in the README I walkthrough having
the user download the Driverless AI MOJO SP, which comes with the MOJO2 Runtime JAR.

Removed the uploaded mojo-pipeline/ folder that was downloaded from Driverless AI. It's no longer
on GitHub.

Removed the uploaded testData/ folder from GitHub. I include instructions on how to get the
data in the README. For batch scoring, the user uses example.csv that comes with mojo-pipeline/
folder. For real-time scoring, the user will run bash commands to break example.csv into smaller bits.

Updated the code in ExecuteDaiMojoScoringPipeline.java (custom NiFi processor)

- on line 168, I use Java's getLogger for proper logging instead of dumping stacks to stderr/stdout
- on line 236, I put the mojoPipelineUUID before the scoring occurs
- on line 255, RecordSetWriter is instantiated as a try resource and it has multilevel inheritance
to AutoCloseable, so I removed the try catch statements that were used to manually close the writer
since when execution leaves the try resource block, the writer is automatically closed.
- Originally the predict() method was doing more than just predicting, so I created a method
called convertRecordToMojoFrame() that translates the NiFi Record to a MojoFrame and I also
created another method called convertMojoFrameToRecord() that translates the MojoFrame to
a NiFi Record. So, now if you look at lines 238 to 245 or lines 265 to 270, you will see the
NiFi record is converted to a MojoFrame, the model is executed on the MojoFrame to do scoring,
and the scored MojoFrame is converted to a NiFi scored record.

james94 committed:
Updated the code so now there is only 1 call to predict()

Before this update, the code had 2 different calls to the predict() method. The first predict() call was
performed on the first NiFi record. The reason for doing this was to get the write schema from the scored
NiFi record since NiFi doesn't know the write schema ahead of time. However, a better approach with the help
of the Driverless AI MOJO Pipeline model was to get the scored MojoFrame Meta Data from the model since it
knows the scored schema ahead of time. The development involved with adding this update required converting
from DAI MojoFrame Meta Data to NiFi Record Schema since NiFi doesn't understand DAI MojoFrame Meta Data. In
the long run, getting the MojoFrame Meta Data from the model would save of some execution time since it only
needs to be done once after we load the MojoPipeline model. Then later in the code we pass the scored
MojoRecordSchema to NiFi's writerFactory.getSchema(...) method to create the scored writeSchema.

orendain reviewed (commented):
Wanted to comment on what I got around to looking at this morning, but can provide some more feedback in a bit.

orendain commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+	@SuppressWarnings("unchecked")
+	private MojoFrame convertRecordToMojoFrame(final Record record) {
+		// Convert record to hash map
+		Map<String, Object> recordMap = (Map<String, Object>) DataTypeUtils.convertRecordFieldtoObject(record, RecordFieldType.RECORD.getRecordDataType(record.getSchema()));
```
Separating this into 2 lines would (IMHO) increase readability and thus maintainability:
```
DataType dataType = RecordFieldType.RECORD.getRecordDataType(record.getSchema());
Map<String, Object> recordMap = (Map<String, Object>) DataTypeUtils.convertRecordFieldtoObject(record, dataType);
```

james94 replied:
@orendain thank you for your recommendation. I applied the update and I can see it makes a huge difference in readability.

orendain commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+		// Convert record to hash map
+		Map<String, Object> recordMap = (Map<String, Object>) DataTypeUtils.convertRecordFieldtoObject(record, RecordFieldType.RECORD.getRecordDataType(record.getSchema()));
+		
+		// Get an instance of a MojoFrameBuilder that will be used to make a MojoFrame
+		MojoFrameBuilder frameBuilder = model.getInputFrameBuilder();
+		// Get an instance of a MojoRowBuilder that will be used to construct a row for this builder
+		MojoRowBuilder rowBuilder = frameBuilder.getMojoRowBuilder();
```
Already commented on something similar, but just want to point out that this block of code is self-documenting and comments are thus unnecessary.  This is due to the great work by the developers of the Mojo2 library - another reason why nicely-named functions are important - because downstream code is easy to read 😃!

I have learned that code comments are often cumbersome for developers (it takes too long to read) and often try to explain what convoluted code can't do on its own.  Since your code is so clean - removing the comments would make this function more concise and to the point.

james94 replied:
@orendain thanks for sharing your insight. I removed the comments and freed up some space. The method looks much easier to read too.

orendain commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+		// Append a row from the current state of the rowBuilder
+		frameBuilder.addRow(rowBuilder);
+		
+		// Construct a MojoFrame
+		MojoFrame iframe = frameBuilder.toMojoFrame();
+		
+		return iframe;
```
`return frameBuilder.toMojoFrame();` (without comment) takes up 75% less space and is just as clear as the above.

james94 replied:
@orendain great point. I can see the difference in how much more concise this part of the method has gotten.

orendain commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+			// Cast Object array to appropriate Data Type
+			switch (mojoFrameColDataType) {
+			case Bool:
+				boolean[] columnBoolArray = (boolean[]) mojoFrameColumnData;
```
I'd consider moving this entire switch statement into a function of its own, maybe returning `column*Array[...]`.  It would better encapsulate the logic, and make this loop much shorter and easier to read.

james94 replied:
@orendain thanks for the suggestion. I was actually thinking about how I would move the switch statement over to a method. I applied the update and it will be in the new commit. The code in `convertMojoFrameToRecord` is easier to read now and more concise too.

orendain commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+				break;
+			}
+		}
+		final Record predictedRecord = DataTypeUtils.toRecord(recordMap, "r");
```
Some of your local variables are being marked as `final`, while others aren't.  _In my experience_ (and in no way an indication of best practices), variables are either only marked `final` when they gain a clear benefit, or variables are marked `final` at all times unless they must be mutable.

In other words: I'd recommend some consistency.  There's no "best" approach IMHO - as long as one is followed consistently.

james94 replied:
@orendain thanks for your insight. I decided to get rid of this `final` on line 428 and return `DataTypeUtils.toRecord(recordMap, "r")` from the method.

orendain reviewed (changes requested):
Marking this as `Request changes` only because of the issue with `scoredColumnDataType` where they may be an issue with the logic of the function.

james94 commented on mojo-nifi/README.md:
```diff
@@ -0,0 +1,310 @@
+# remove file's 1st line, the header
+echo -e "$(sed '1d' example.csv)\n" > example.csv
+
+# split example.csv into smaller real-time files each with 1 line
+split -l 1 example.csv test_
+
+# add .csv extension to all test_* files in folder
+for f in test* ; do mv "$f" "${f}.csv"; done
```
@orendain thanks for the suggestion. I updated split command to use -additional-suffix and numeric suffix. So now instead of 5 lines, only 1 line is used for the command and 1 line for the comment.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-nar/pom.xml:
```diff
@@ -0,0 +1,47 @@
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
```
@orendain these files were automatically licensed under Apache License 2.0 when I generated the NiFi Quickstart Processor project. I removed the license for each file that had one. When a user runs mvn clean install, a single LICENSE file is generated in a target directly. Thank you for sharing your observation.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/pom.xml:
```diff
@@ -0,0 +1,95 @@
+    <parent>
+        <groupId>org.apache.nifi</groupId>
+        <artifactId>nifi-h2o-record-bundle</artifactId>
+        <version>1.11.4</version>
```
@orendain thanks for your suggestion. When I generated the NiFi Quickstart Processor project, it came with the nifi version numbers as constants. I updated the pom.xml files to use maven variable `nifi.version` in almost all places that had a hardcoded nifi version number. The exception to this update was to the places where there is a nifi version hardcoded in between the `<parent>` tags and `<groupId>...</package> tags. That way the compiler did not complain.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+	@WritesAttribute(attribute = "record.count", description = "The number of records in an outgoing FlowFile"),
+	@WritesAttribute(attribute = "mime.type", description = "The MIME Type that the configured Record Writer indicates is appropriate"),
+})
+@CapabilityDescription("Executes H2O's Driverless AI MOJO Scoring Pipeline in Java Runtime to do batch "
```
@orendain thank you for the suggestion. I made the update and it will be in the new commit.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+		_properties.add(RECORD_READER);
+		_properties.add(RECORD_WRITER);
+		_properties.add(MODULE);
+		_properties.add(PIPELINE_MOJO_FILEPATH);
+		properties = Collections.unmodifiableList(_properties);
+		
+		final Set<Relationship> _relationships = new HashSet<>();
+		_relationships.add(REL_SUCCESS);
```
@orendain thanks for the suggestion. I updated the code for `_properties` to be `propertiesBuilder` and `_relationships` to be `relationshipsBuilder`. It will be in the new commit.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+	
+	@OnScheduled
+	public void onScheduled(final ProcessContext context) {
+		// gets called once when the processor is scheduled to run
```
@orendain thanks for the recommendation. I changed the single comment to multi comment. It will be in the new commit.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+		// gets called once when the processor is scheduled to run
+		
+		try {
+			// Load Mojo Pipeline model (includes feature engineering + ML model)
+			final String pipelineMojoPath = context.getProperty(PIPELINE_MOJO_FILEPATH).getValue();
+			model = MojoPipeline.loadFrom(pipelineMojoPath);
+			
+			// Get Mojo Pipeline model UUID
+			mojoPipelineUUID = "pipeline.mojo uuid " + model.getUuid();
+			
+			// Convert MojoFrameMeta to NiFi RecordSchema
+			scoredMojoRecordSchema = convertMojoFrameMetaToRecordSchema();
```
@orendain thanks for your suggestion. I have removed comments for variables and methods that are self-documenting. It will be in the new commit.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+			scoredMojoRecordSchema = convertMojoFrameMetaToRecordSchema();
+
+		} catch (Exception e) {
+			getLogger().error("Unable to load MOJO from pipelineMojoPath or " + "Unable to get MOJO Pipeline Model UUID or " +
```
@orendain for dealing with the exceptions that come from `onScheduled`, I decided to use throws to indicate the type of exceptions `IOException` or `LicenseException` that could occur while the code is running in this method.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+
+			if (scoredColumnDataType == null) {
+				throw new ProcessException("Error converting MojoFrame Column Data Type to NiFi Data Type");
+			}
```
@orendain I updated the code based on your suggestion. I moved the switch statement to method `convertMojoTypeToRecordType`. I declared `recordType` of DataType, but then when I returned `recordType`, IDE complained about `recordType` not being initialized, so I initialized the object to `null`. In the default case, I `throw ProcessException` since conversion failed. However, if conversion succeeds, I return `recordType` to `scoredColumnDataType` as you suggest above. This update will be in new commit.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+		final List<ValidationResult> results = new ArrayList<>(super.customValidate(validationContext));
+		final String pipelineMojoPath = validationContext.getProperty(PIPELINE_MOJO_FILEPATH).isSet() ? validationContext.getProperty(PIPELINE_MOJO_FILEPATH).getValue() : null;
+		
+		if(pipelineMojoPath == null) {
```
@orendain thanks for this suggestion. I updated the code to check if the PIPELINE_MOJO_FILEPATH is not set. I saved a line of code and also the condition looks more readable.

james94 replied:
@orendain I decided to remove customValidate. Now I am using a NiFi StandardValidator called `StandardValidators.FILE_EXISTS_VALIDATOR` for validating that the file for `PIPELINE_MOJO_FILEPATH` exists. If it doesn't exist, NiFi UI will show an alert message on the top left corner of the `ExecuteDaiMojoScoringPipeline` processor and won't let the user start the processor until they add a correct file path.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+		FlowFile scored = null; // flowfile contents contains scored (predicted) data
+		
+		try (final InputStream in = session.read(original);
+			 final RecordReader reader = readerFactory.createRecordReader(original, in, getLogger())
```
@orendain thanks for catching that. I replaced `getLogger()` with `logger`.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+	}
+	
+	private MojoFrame predict(MojoFrame mojoFrame, MojoPipeline model) {
+		// Executes the pipeline of transformers as stated in this model's Mojo file
+		// In other words, use Mojo to predict the labels from iframe real world data
```
@orendain thank you for the suggestion. I added a multiline comment above the method and now it is also easier to read.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/java/org/apache/nifi/processors/h2o/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,431 @@
+				session.getProvenanceReporter().modifyContent(scored, "Modified With " + mojoPipelineUUID, stopWatch.getElapsed(TimeUnit.MILLISECONDS));
+				logger.debug("Scored {}", new Object[] {original});
+			}
+		} catch (final Exception ex) {
```
@orendain after doing some inspecting on the type of exceptions that can be thrown in the try block, I narrowed down exceptions to be `IOException | MalformedRecordException | SchemaNotFoundException`. I also removed `final` from that catch statement. I don't think it is important. catching these more specific exceptions will be in the new commit

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/test/java/org/apache/nifi/processors/h2o/record/TestExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,304 @@
+	}
+	
+	@Before
+	public void setup() throws Exception {
```
@orendain great point. I added more specific exceptions `InitializationException, IOException` for throws. It will be in the new commit.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/test/java/org/apache/nifi/processors/h2o/record/TestExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,304 @@
+		try {
+			runner.addControllerService("writer", writerService);
+		} catch(InitializationException e) {
+			throw new IOException(e);
```
@orendain that is true. I just removed the try catch blocks and that method is much easier to read. Thanks for the suggestion.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/test/java/org/apache/nifi/processors/h2o/record/TestExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,304 @@
+		}
+	}
+	
+	private float randFloat(float min, float max) {
```
@orendain thanks for the recommendation. I just added that update and it will be in the new commit.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/test/java/org/apache/nifi/processors/h2o/record/TestExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,304 @@
+			
+			readerService.addSchemaField("hydraulicTestData", RecordFieldType.RECORD);
+			
+			Random r = new Random();
```
@orendain you are correct. It turns out I didn't need this variable. It is removed.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/test/java/org/apache/nifi/processors/h2o/record/TestExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,304 @@
+		runner.assertNotValid();
+	}
+	
+	private void generateHydraulicTestData(int numRecords, final BiFunction<Integer, MockRecordParser, Void> recordGenerator) {
+		
+		// based on hydraulicTestData.avsc file
+		if(recordGenerator == null) {
```
@orendain thank you for the suggestion. I created a helper method `generateDefaultHydraulicTestData(int numRecords)` that calls `generateHydraulicTestData(numRecords, null)` encapsulating the `null` value.

james94 commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/test/java/org/apache/nifi/processors/h2o/record/TestExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,304 @@
+	}
+	
+	@Test
+	public void testRelationshipsCreated() throws IOException {
```
From the code I have seen from the NiFi team, the tests are usually `public`. For example,

- [TestGetFile.java](https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestGetFile.java)
- [TestPutFile.java](https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFile.java)

The name convention for the function `test<Description>()`, I have seen in a lot of NiFi test java programs.

Yeah that is an interesting observation you have seen with most Java test programs. For NiFi, most of the Java tests I have seen start with `test`, then the description and the test methods are marked as public. I will let you know what I notice with Apache Flink. There convention for writing Java tests may be similar or different.

james94 commented:
@nkpng2k The 2 mojo runtime jars in the pr have been removed. Maven has been used to manage the dependencies. Now Maven downloads the `mojo2-runtime-api` and `mojo2-runtime-impl`.

james94 commented:
@orendain For `scoredColumnDataType` inside `convertMojoFrameMetaToRecordSchema()`, I updated the logic based on your suggestion for resetting the `recordType = null` at the beginning of each loop iteration, then based on the case chosen in switch case for the `mojoType`, I return the appropriate `recordType` from `convertMojoTypeToRecordType(MojoColumn.Type mojoType)` to `convertMojoFrameMetaToRecordSchema()`, which eventually returns the scored RecordSchema. I will submit the latest commit soon, so you can see the updates.

All I have left for now is to break up the long `onTrigger()` method into multiple small methods that make `onTrigger()` easier to follow. I could work on that update later.

james94 committed:
Updated DAI MOJO NiFi Deployment based on Nick & Edgar's Feedback

- In Maven pom.xml, added mojo2-runtime-api and mojo2-runtime-impl for Maven to manage dependencies
- In mojo-nifi/README.md, updated split command to use numeric suffix and additional suffix
- In files that had Apache License 2.0 comments, removed them since one single License is autogenerated
when user builds project with mvn clean install
- In pom.xml, updated all constant nifi versions with Maven nifi.version variable
- In Java both java programs, declared more specific exception types
- In Java test program, updated setup method to indicate that certain Exceptions may be thrown
- In Java test program, added static to randFloat
- In Java test program, removed unused Random r variable
- In Java test program, added helper generateDefaultHydraulicTestData(numRecords) method to encapsulate null
- In NiFi Java test programs, it looks like they mark their tests as public with name convention test<Description()>
- In Java ExecuteDaiMojoScoringPipeline, changed _properties to propertiesBuilder and did similar for other variables
that started with underscores
- Also added comments above methods for further explaining method if it was needed
- For code that was self-documenting, removed comments
- For convertMojoFrameMetaToRecordSchema() method and convertMojoTypeToRecordType(MojoColumn.Type mojoType), made sure
to deal with null based on suggestion, so now the logic should be correct. As you can see also moved the switch statement
to convertMojoTypeToRecordType(MojoColumn.Type mojoType)
- Removed customValidate method since NiFi comes with plenty of StandardValidators. Used a NiFi validator to verify that the
file for the pipeline mojo filepath property exists
- In most places, changed getLogger() to logger
- At the last catch(...) in onTrigger() method, added more specific Exceptions to catch and removed final keyword from line 327
- For recordMap, separated it into 2 lines for easier readability: initialize dataType, then initialize recordMap with record
and dataType
- For switch case statement that was checking mojoFrameColDataType and then converting mojoColumnData to appropriate dataType,

moved it to getMojoFrameDataEntry(...)

orendain commented on mojo-nifi/nifi-nar-bundles/nifi-daimojo-record-bundle/nifi-daimojo-record-processors/src/main/java/org/apache/nifi/processors/daimojo/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,417 @@
+	 * @return
+	 */
+	private DataType convertMojoTypeToRecordType(MojoColumn.Type mojoType) {
+		DataType recordType = null;
+		
+		switch (mojoType) {
+		case Bool:
+			recordType = RecordFieldType.BOOLEAN.getDataType();
+			break;
+		case Int32:
+			recordType = RecordFieldType.INT.getDataType();
+			break;
+		case Int64:
+			recordType = RecordFieldType.LONG.getDataType();
+			break;
+		case Float32:
+			recordType = RecordFieldType.FLOAT.getDataType();
+			break;
+		case Float64:
+			recordType = RecordFieldType.DOUBLE.getDataType();
+			break;
+		case Str:
+			recordType = RecordFieldType.STRING.getDataType();
+			break;
+		default:
+			throw new ProcessException("Error converting Mojo Column Type to NiFi Data Type");
+		}
+		return recordType;
```
nit: You could save quite a few lines by returning immediately instead of storing the data type to a field.

I.e.:
```
case Int32:
    return RecordFieldType.INT.getDataType();
```

You'd then have no need for breaks, or the `DataType` variable.

orendain commented on mojo-nifi/nifi-nar-bundles/nifi-daimojo-record-bundle/nifi-daimojo-record-processors/src/main/java/org/apache/nifi/processors/daimojo/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,417 @@
+	public void onScheduled(final ProcessContext context) {
+		final String pipelineMojoPath = context.getProperty(PIPELINE_MOJO_FILEPATH).getValue();
+
+			try {
+				model = MojoPipelineService.loadPipeline(new File(pipelineMojoPath));
+				
+				mojoPipelineUUID = "pipeline.mojo uuid " + model.getUuid();
+			} catch (LicenseException | IOException e) {
+				getLogger().error("Unable to load MOJO from pipelineMojoPath due to " + e.toString());
+			}
```
nit: Over indented.

Also, in addition to (or instead of) logging an error, you may want to not catch the exception, instead letting it bubble up back to whatever NiFi code calls this function.

According to the following, throwing an exception here will both present the NiFi user with an error and stop the rest of the processors functions from being called.  https://docs.cloudera.com/HDPDocuments/HDF3/HDF-3.5.1/adding-functionality-to-apache-nifi/content/onscheduled.html

orendain commented on mojo-nifi/nifi-nar-bundles/nifi-daimojo-record-bundle/nifi-daimojo-record-processors/src/main/java/org/apache/nifi/processors/daimojo/record/ExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,417 @@
+	 * @return
+	 */
+	private Object getMojoFrameDataEntry(MojoColumn.Type mojoType, Object mojoFrameData) {
+		Object mojoFrameDataEntry = null;
+		
+		switch (mojoType) {
+		case Bool:
+			boolean[] columnBoolArray = (boolean[]) mojoFrameData;
+			mojoFrameDataEntry = columnBoolArray[0]; // since there is only one row in a MojoFrame, we apply static 0
+			break;
+		case Int32:
+			int[] columnIntArray = (int[]) mojoFrameData;
+			mojoFrameDataEntry = columnIntArray[0];
+			break;
+		case Int64:
+			long[] columnLongArray = (long[]) mojoFrameData;
+			mojoFrameDataEntry = columnLongArray[0];
+			break;
+		case Float32:
+			float[] columnFloatArray = (float[]) mojoFrameData;
+			mojoFrameDataEntry = columnFloatArray[0];
+			break;
+		case Float64:
+			double[] columnDoubleArray = (double[]) mojoFrameData;
+			mojoFrameDataEntry = columnDoubleArray[0];
+			break;
+		case Str:
+			String[] columnStringArray = (String[]) mojoFrameData;
+			mojoFrameDataEntry = columnStringArray[0];
+			break;
+		default:
+			throw new ProcessException("Unable to convert MojoFrame Data Object to Data Type to get Data Entry");
+		}
+		return mojoFrameDataEntry;
```
Same comment as above: consider returning instead of saving the a variable.  It'll save you maybe a dozen lines of code and reduce complexity of the function by reducing the use of a variable.

orendain commented on mojo-nifi/nifi-nar-bundles/nifi-daimojo-record-bundle/nifi-daimojo-record-processors/src/test/java/org/apache/nifi/processors/daimojo/record/TestExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,239 @@
+import org.apache.nifi.csv.CSVRecordSetWriter;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.reporting.InitializationException;
+//import org.apache.nifi.schema.access.SchemaAccessUtils;
```
tip:  Most IDEs have some sort of "Optimize Imports" feature, where it will remove all unused imports.  I tend to hit both my `Optimize Imports` and `Auto Format` shortcuts right before I save/commit, to automatically clean up my code.

orendain commented on mojo-nifi/nifi-nar-bundles/nifi-daimojo-record-bundle/nifi-daimojo-record-processors/src/test/java/org/apache/nifi/processors/daimojo/record/TestExecuteDaiMojoScoringPipeline.java:
```diff
@@ -0,0 +1,239 @@
+	
+	private void generateDefaultHydraulicTestData(int numRecords) {
+		generateHydraulicTestData(numRecords, null);
+	}
```
I think we already discussed this offline, but dropping this marker here just to point out where the use of `null` could be removed.

orendain reviewed (approved):
LGTM.  Just added some formatting and style comments.

I believe we had gone through this offline, but in case we hadn't then apologies on the delay.

james94 committed:
Merge branch 'master' into mojo-nifi

james94 committed:
Updated links and DAI MOJO SP version in README

Updated outdated links for dai-deployment-examples repo from forked to h2oai one.

Updated NiFi download link to the archived one since the previous no longer is supported.

Updated test environment related info for DAI version to be 1.9.0 for that was the one that the MOJO SP was built on. Updated MOJO SP version to be 2.4.8 since that was the Java Runtime API version used in building this custom NiFi Processor ExecuteDaiMojoScoringPipeline.

james94 committed:
Update pom to pull latest DAI MOJO SP API version

Updated pom version dependency for daimojo.version to be 2.4.8 since that was the latest version of the DAI MOJO Scoring Pipeline pipeline.mojo downloaded from DAI Experiment tested with this processor. Also tested this processor with this DAI MOJO SP API version when compiling processor during Cloudera Certification testing.

james94 commented:
@orendain I am going to merge this PR with the updates made to this custom NiFi processor so far. I believe many of the critical issues have been addressed from the feedback. There is always room for improvement. I will leave it in the good hands of the open source for future contributors who may see areas of improvement. Also Edgar has pointed out some more areas of improvement that I figure we could incorporate in the next PR, but for now I am happy with what has been addressed so far for this PR and will merge.
